src/databricks/labs/ucx/assessment/dashboards.py

```diff
@@ -0,0 +1,370 @@
+                continue
+            query_ids.append(widget.visualization.query.id)
+        return cls(
+            id=dashboard.id or cls.id,
+            name=dashboard.name or cls.name,
+            parent=dashboard.parent or cls.parent,
+            query_ids=query_ids,
+            tags=dashboard.tags or [],
+        )
+
+    @classmethod
+    def from_sdk_lakeview_dashboard(cls, dashboard: SdkLakeviewDashboard) -> Dashboard:
+        assert dashboard.dashboard_id
+        lsql_dashboard = _convert_sdk_to_lsql_lakeview_dashboard(dashboard)
+        query_ids = [dataset.name for dataset in lsql_dashboard.datasets]
+        return cls(
+            id=dashboard.dashboard_id,
+            name=dashboard.display_name or cls.name,
+            parent=dashboard.parent_path or cls.parent,
+            query_ids=query_ids,
+        )
+
+
+class RedashDashboardCrawler(CrawlerBase[Dashboard]):
+    """Crawler for Redash dashboards."""
+
+    def __init__(
+        self,
+        ws: WorkspaceClient,
+        sql_backend: SqlBackend,
+        schema: str,
+        *,
+        include_dashboard_ids: list[str] | None = None,
+        include_query_ids: list[str] | None = None,
+        debug_listing_upper_limit: int | None = None,
+    ):
+        super().__init__(sql_backend, "hive_metastore", schema, "redash_dashboards", Dashboard)
+        self._ws = ws
+        self._include_dashboard_ids = include_dashboard_ids or []
+        self._include_query_ids = include_query_ids or []
+        self._debug_listing_upper_limit = debug_listing_upper_limit
+
+    def _crawl(self) -> Iterable[Dashboard]:
+        dashboards = []
+        for sdk_dashboard in self._list_dashboards():
+            if sdk_dashboard.id is None:
+                continue
+            dashboard = Dashboard.from_sdk_redash_dashboard(sdk_dashboard)
+            dashboards.append(dashboard)
+        return dashboards
+
+    def _list_dashboards(self) -> list[SdkRedashDashboard]:
+        if self._include_dashboard_ids:
+            return self._get_dashboards(*self._include_dashboard_ids)
+        try:
+            dashboards_iterator = self._ws.dashboards.list()
+        except DatabricksError as e:
+            logger.warning("Cannot list Redash dashboards", exc_info=e)
+            return []
+        dashboards: list[SdkRedashDashboard] = []
+        # Redash APIs are very slow to paginate, especially for large number of dashboards, so we limit the listing
+        # to a small number of items in debug mode for the assessment workflow just to complete.
+        while self._debug_listing_upper_limit is None or self._debug_listing_upper_limit > len(dashboards):
+            try:
+                dashboards.append(next(dashboards_iterator))
+            except StopIteration:
+                break
+            except DatabricksError as e:
+                logger.warning("Cannot list next Redash dashboards page", exc_info=e)
+                break
+        return dashboards
```

Comment:
I think this can be simplified somewhat:
 - SDK calls to `.list()` normally allow a maximum number of results.
 - `itertools.islice()` can be used to ignore everything beyond the first $$n$$ results.

That would lead to something like:
```python
try:
    if self._debug_listing_upper_limit is None:
        dashboards_iterator = self._ws.dashboards.list()
    else:
        dashboards_iterator = itertools.islice(self._ws.dashboards.list(page_size=self._debug_listing_upper_limit), self._debug_listing_upper_limit)  # might also need page=0 or page=1 or something.
    # Force SDK errors to be materialised here rather than deferred.
    return list(dashboards_iterator)
except DatabricksError as e:
    logger.warning("Cannot list Redash dashboards", exc_info=e)
    return []
```
This also eliminates handling the same error twice.

Reply:
Though, I like the use of `islice` it does not reach the same functionality, see `test_redash_dashboard_crawler_handles_databricks_error_on_iterate`. If we get a `TooManyRequests` half way listing, it returns the dashboards listed until then with the current code, with your seggestion it returns no dashboards

Reply:
That's a good concern, but I don't think it can happen? There's only a single call being made (`.list()`) which will either succeed with (a page of) results or fail immediately.

I don't think that unit test emulates an error that can occur; the only point that `TooManyRequests` can be raised is immediately upon receipt of the HTTP request, _before_ the first dashboard is returned. (Due to the way coroutines work raising the error is probably deferred until the _first_ dashboard is yielded, but after that I don't think it can happen and the consequence is the same: an empty list.)

Thinking about it some more… arguably the crawler for this particular case should also use a retry-handler: there's something to be said for returning partial results, but if the `.list()` itself fails, especially due to `TooManyRequests`, then it should retry after the delay including in the exception. I'm also a little uncomfortable about just returning an empty list: under the current code we've had a failure that prevents it from working properly but the `assessment` will succeed. What do you think about this?

Reply:
yeah, the difficulty here is that I agree, but want to code defensively and avoid too many assumptions about the API inner working.

On the single `list`, it might call the API multiple times, see https://github.com/databricks/databricks-sdk-py/blob/980166030c688e08a52d0f55389abd766f1d928d/databricks/sdk/service/dashboards.py#L1350-L1357 

I expect if we set the page size to the debug limit, then it should call API just once, but I am not 100% certain

Reply:
I am leaning to keeping the code as it is

Reply:
we could log the error as an `logger.error` to make sure the assessment workflow fails

Reply:
#3440

Reply:
And promoted log message to ERROR level